### PR TITLE
Fix #2083 - [BUG] FrameLimiter Sleep strategy sleeps too long

### LINKED ISSUE #2083: [BUG] FrameLimiter Sleep strategy sleeps too long
I think there is an error in the FrameLimiter implementation logic for the `Sleep` strategy, unless I'm misunderstanding something.

The docs say (emphasis mine):
> Sleep will call `thread::sleep` with a **duration of 0 milliseconds** until the frame duration has passed.

In the code, the `Sleep` strategy is called with a zero duration:
```rust
const ZERO: Duration = Duration::from_millis(0);
....
Sleep => self.do_sleep(ZERO)
```

Working through the `do_sleep` function (comments mine, durations expressed in milis):
```rust
fn do_sleep(&self, stop_on_remaining: Duration) {    // stop_on_remaining = 0 milis
    let frame_duration = self.frame_duration - stop_on_remaining;    // assume 16ms frame, frame_duration = 16 - 0 = 16 milis
    loop {
        let elapsed = Instant::now() - self.last_call;    // assume last_call 4ms, elapsed = 4 milis
        if elapsed >= frame_duration {    // if 4 > = 16 is false
            break;
        } else {    // so we end up here
            sleep(frame_duration - elapsed);    // sleep(16 - 4) | sleep(12)
        }
    }
}
```

I think the error here is that the `stop_on_remaining` parameter is being used for two different things:
1) For the `SleepAndYield` strategy, setting the time before the next frame that we must stop sleeping and start yielding.  This - I think - works.
2) For the `Sleep` strategy, setting the time we want to sleep for, each time we call sleep.

I see two possible solutions:
a) add another parameter to the `do_sleep` method.  Something like `max_sleep_length`.  This might be useful for the `SleepAndYield` strategy too since it would remove the possibility of very long sleeps (and thread unresponsiveness)
b) use a separate method or loop for the `Sleep` strategy.

I don't know this code well, and I'm finding the names a little obtuse, so please tell me if I've got this wrong.

Thanks

### PR DESCRIPTION
**Pull request by borup3**
## Description

This PR is a proposed fix for issue https://github.com/amethyst/amethyst/issues/2083 that has observed an apparent problem with `FrameLimiter`'s implementation of `FrameRateLimitStrategy::Sleep`. As mentioned in the issue, this is highly likely to over-sleep due to the imprecise nature of operating system sleeps. The proposed fix is a halfway house between sleeping and yielding, not unlike `FrameRateLimitStrategy::SleepAndYield(Duration)` did but with an automatic duration equal to half the requested frame time.

The primary reasoning behind this change is to reduce the complexity of the struct for end-users who are unfamiliar with the undesirable intricacies of operating system sleeps, from a game development point of view. These users may consider `FrameRateLimitStrategy::Sleep` a sensible choice due to its labelling as the more pro battery-life approach and may not understand why their game may suffer from inconsistent timing issues as a result. This fix attempts to strike a balance between `Sleep` and `Yield` without requiring the user to be aware of the intricacies of sleeping, so that `Sleep` becomes a sensible default go-to option.

`FrameRateLimitStrategy::Sleep` has been removed. It is now called `FrameRateLimitStrategy::SleepAndYield`. The prior `FrameRateLimitStrategy::SleepAndYield` is now called `FrameRateLimitStrategy::SleepAndYieldDuration`.

The new `FrameRateLimitStrategy::SleepAndYield` implementation will attempt to sleep for the first 50% of the desired frame time and yield for the remaining 50%. This is done automatically for the end-user of the struct when they invoke `set_rate`. The new `FrameRateLimitStrategy::SleepAndYield` should now offer a sensible default behavior in place of `FrameRateLimitStrategy::Yield`, albeit this change has not been made in this PR.

This demotes `FrameRateLimitStrategy::SleepAndYieldDuration(Duration)` to being nothing more than an alternate version of `FrameRateLimitStrategy::SleepAndYield` that grants the end-user more control over the sleep-to-yield ratio should they need it.

## Additions

This PR adds a new function `pub fn FrameLimiter::get_frame_duration(&self) -> Duration` that returns `self.frame_duration`.

## Removals

`FrameRateLimitStrategy::Sleep` has been removed. It is now called `FrameRateLimitStrategy::SleepAndYield`. The prior `FrameRateLimitStrategy::SleepAndYield` is now called `FrameRateLimitStrategy::SleepAndYieldDuration`.

## Modifications

The `FrameLimiter` struct has a new member `sleep_barrier` that is assigned in `FrameLimiter::set_rate` to half of the desired frame duration. The implementation of `FrameRateLimitStrategy::SleepAndYield` in `FrameLimiter::wait` has been changed to yield once the `sleep_barrier` has been breached.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Added unit tests for new code added in this PR.
- [X] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.
- [X] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Updated the content of the book if this PR would make the book outdated.

If this modified or created any rs files:

- [X] Ran `cargo +stable fmt --all`
- [X] Ran `cargo clippy --workspace --features "empty"` (may require `cargo clean` before)
- [X] Ran `cargo build --features "empty"`
- [X] Ran `cargo test --workspace --features "empty"`

### CODE REVIEW

**Comment by 0x6273** [comment]
I think this makes the names a bit confusing. If there is one enum named `SleepAndYield`, I would expect the enum named `Sleep` to just sleep, and not yield.
Why not get rid of `Sleep` and make `SleepAndYield` default to sleeping for half a frame?

**Comment by borup3** [comment]
> I think this makes the names a bit confusing. If there is one enum named `SleepAndYield`, I would expect the enum named `Sleep` to just sleep, and not yield.
> Why not get rid of `Sleep` and make `SleepAndYield` default to sleeping for half a frame?

I think since the dominant behavior of `Sleep` is still sleeping the name still makes sense. Whereas for `SleepAndYield` you can customize the dominant behavior to your needs. For games, you should never limit frames purely with sleeps so it shouldn't even be offered as an option. You don't want to take away the users ability to take full control over their limiter (with `SleepAndYield`) if they want that for whatever reason.

This approach is backwards-compatible and gives any games that are currently built with `Sleep` a better solution without needing any effort or awareness on their end to update their code. Otherwise I would recommend naming `Sleep` as `SleepAndYield` and changing `SleepAndYield` to `SleepAndYieldDuration`. And not offer a `Sleep` behavior at all.

**Comment by 0x6273** [comment]
> Otherwise I would recommend naming `Sleep` as `SleepAndYield` and changing `SleepAndYield` to `SleepAndYieldDuration`. And not offer a `Sleep` behavior at all.

That makes a lot more sense to me, since they both sleep and yield, and the only difference is that the former is hard-coded to sleep for half a frame duration. The breakage is probably not a big deal, since most people just use the default, and it's a trivial fix for those that don't.

**Comment by 0x6273** [comment]
If you really don't want to break people's code, you could slap a `#[deprecated]` on `Sleep`, and name the new frame limiter `SleepAndYieldHalf`.

**Comment by borup3** [comment]
I've renamed Sleep to SleepAndYield and SleepAndYield to SleepAndYieldDuration.

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,8 @@ The format is based on [Keep a Changelog][kc], and this project adheres to
 
 ### Fixed
 
+[#2083]: https://github.com/amethyst/amethyst/pull/2479
```
Comment (0x6273) [inline docs/CHANGELOG.md:19]:
```suggestion
[#2479]: https://github.com/amethyst/amethyst/pull/2479
```
Also, you need a line above it like: `- Short summary of the fix ([#2479])`

**Review by 0x6273 — APPROVED** [review]

**Comment by jojolepro** [comment]
I think it would be preferable to eventually move the sleeping to an external crate. One like this: https://crates.io/crates/spin_sleep
Currently, spin_sleep doesn't support wasm (no access to the `Time` std api). Once it does, delegating the sleeping behavior to a library would allow us to focus on maintaining other parts of the engine instead of dealing over and over with sleep/timing/cpu usage issues. This needs futher discussion with other members so we can agree on a strategy going forward. As such, I opened a forum discussion here: https://community.amethyst.rs/t/sleep-and-cpu-usage-going-forward/1763

**Comment by jojolepro** [comment]
Following conversations, we will close this PR and instead switch to the spin_sleep crate. Thanks for your contribution!

**Pull request closed without merging**